src/MongoDB.Driver.Core/Core/Operations/ChangeStreamCursor.cs

```diff
@@ -81,20 +81,18 @@ public void Dispose()
         public bool MoveNext(CancellationToken cancellationToken = default(CancellationToken))
         {
             bool hasMore;
-            try
+            while (true)
             {
-                hasMore = _cursor.MoveNext(cancellationToken);
-            }
-            catch (Exception ex)
-            {
-                if (RetryabilityHelper.IsResumableChangeStreamException(ex))
+                try
                 {
-                    _cursor = _changeStreamOperation.Resume(_binding, cancellationToken);
                     hasMore = _cursor.MoveNext(cancellationToken);
+                    break;
                 }
-                else
+                catch (Exception ex) when (RetryabilityHelper.IsResumableChangeStreamException(ex))
                 {
-                    throw;
+                    var newCursor = _changeStreamOperation.Resume(_binding, cancellationToken);
+                    _cursor.Dispose();
+                    _cursor = newCursor;
```

Comment:
Don't replace the _cursor until the newCursor has been created without throwing an exception.

As a separate minor issue we were neglecting to call Dispose on the old cursor before.